python310Packages.bokeh: Build from source & enable tests

mweinelt opened the pull request:
To enable the tests we have to fetch the sources from Git, which in turn requires us to build the static files from source.

Bokeh uses setuptools-git-versioning, which has no version override, so we also fix the broken fallback version in the package metadata.

Still some failing tests towards and each iteration takes 10-12 minutes and it seems like at some point the selenium stuff just breaks down.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
python310Packages.bokeh: 2.4.3 -> 3.2.0

https://docs.bokeh.org/en/latest/docs/releases.html#release-3-2-0

Changes the build to a source build and enables tests.

To enable the tests we have to fetch the sources from Git, which in turn
requires us to build the static files from source.

Bokeh uses setuptools-git-versioning, which has no version override, so
we also fix the broken fallback version in the package metadata.

We skip sampledata tests, because that would require an additional
download, and we skip selenium tests because they are crashing with
chromium/chromedriver and getting out of sync with geckodriver/firefox.

dotlambda commented:
Something similar would be nice for `python3Packages.panel`.

mweinelt commented on pkgs/development/python-modules/bokeh/default.nix:
```diff
@@ -51,55 +96,102 @@ buildPythonPackage rec {
+    # FileNotFoundError: [Errno 2] No such file or directory: '/build/source/tests/unit/bokeh/embed/ext_package_no_main/dist/ext_package_no_main.js'
+    "test_with_CDN_resources"
+    "test_with_INLINE_resources"
+    # selenium.common.exceptions.TimeoutException: Message: Failed to read marionette port
+    "test_issue_11694"
+    "test_bkroot_added_to_target"
+    "test_includes_hidden_glyphs_by_default"
```
Basically these exceptions are creeping up towards the end. I can't see what causes them.

mweinelt commented:
> Something similar would be nice for `python3Packages.panel`.

No doubt about that. These changes did cost me some time and energy already, so it's not in the cards for now.

dotlambda commented:
> > Something similar would be nice for `python3Packages.panel`.
> 
> No doubt about that. These changes did cost me some time and energy already, so it's not in the cards for now.

Yeah, this looks like a *lot* of work. I'd hope that panel is set up somewhat similar but I totally understand that you don't feel like tackling that now.

dotlambda commented:
Result of `nixpkgs-review pr 215406` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.optuna</li>
    <li>python310Packages.rl-coach</li>
  </ul>
</details>
<details>
  <summary>16 packages failed to build:</summary>
  <ul>
    <li>python310Packages.datashader</li>
    <li>python310Packages.holoviews</li>
    <li>python310Packages.hvplot</li>
    <li>python310Packages.livelossplot</li>
    <li>python310Packages.nbsmoke</li>
    <li>python310Packages.panel</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.textacy</li>
    <li>python310Packages.wandb</li>
    <li>python311Packages.holoviews</li>
    <li>python311Packages.hvplot</li>
    <li>python311Packages.livelossplot</li>
    <li>python311Packages.nbsmoke</li>
    <li>python311Packages.panel</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python310Packages.bokeh</li>
    <li>python311Packages.bokeh</li>
  </ul>
</details>

dotlambda commented:
Result of `nixpkgs-review pr 215406` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.datashader</li>
    <li>python310Packages.optuna</li>
    <li>python310Packages.rl-coach</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.textacy</li>
  </ul>
</details>
<details>
  <summary>10 packages failed to build:</summary>
  <ul>
    <li>python310Packages.holoviews</li>
    <li>python310Packages.hvplot</li>
    <li>python310Packages.livelossplot</li>
    <li>python310Packages.nbsmoke</li>
    <li>python310Packages.panel</li>
    <li>python311Packages.holoviews</li>
    <li>python311Packages.hvplot</li>
    <li>python311Packages.livelossplot</li>
    <li>python311Packages.nbsmoke</li>
    <li>python311Packages.panel</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.bokeh</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.wandb</li>
    <li>python311Packages.bokeh</li>
  </ul>
</details>

dotlambda commented:
Result of `nixpkgs-review pr 215406` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.datashader</li>
    <li>python310Packages.optuna</li>
    <li>python310Packages.rl-coach</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.textacy</li>
  </ul>
</details>
<details>
  <summary>12 packages failed to build:</summary>
  <ul>
    <li>python310Packages.holoviews</li>
    <li>python310Packages.hvplot</li>
    <li>python310Packages.livelossplot</li>
    <li>python310Packages.nbsmoke</li>
    <li>python310Packages.panel</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.wandb</li>
    <li>python311Packages.holoviews</li>
    <li>python311Packages.hvplot</li>
    <li>python311Packages.livelossplot</li>
    <li>python311Packages.nbsmoke</li>
    <li>python311Packages.panel</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python310Packages.bokeh</li>
    <li>python311Packages.bokeh</li>
  </ul>
</details>

dotlambda commented:
Result of `nixpkgs-review pr 215406` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.optuna</li>
    <li>python310Packages.rl-coach</li>
  </ul>
</details>
<details>
  <summary>11 packages failed to build:</summary>
  <ul>
    <li>python310Packages.datashader</li>
    <li>python310Packages.holoviews</li>
    <li>python310Packages.hvplot</li>
    <li>python310Packages.livelossplot</li>
    <li>python310Packages.nbsmoke</li>
    <li>python310Packages.panel</li>
    <li>python311Packages.holoviews</li>
    <li>python311Packages.hvplot</li>
    <li>python311Packages.livelossplot</li>
    <li>python311Packages.nbsmoke</li>
    <li>python311Packages.panel</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>python310Packages.bokeh</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.textacy</li>
    <li>python310Packages.wandb</li>
    <li>python311Packages.bokeh</li>
  </ul>
</details>

mweinelt commented:
https://github.com/holoviz/panel/issues/4097

Panel isn't compatible with fully bokeh 3.0 yet.

mweinelt committed:
WIP: python310Packages.panel: 0.14.4 -> 1.1.0

https://github.com/holoviz/panel/releases/tag/v1.0.0
https://github.com/holoviz/panel/releases/tag/v1.0.1
https://github.com/holoviz/panel/releases/tag/v1.0.2
https://github.com/holoviz/panel/releases/tag/v1.0.3
https://github.com/holoviz/panel/releases/tag/v1.0.4
https://github.com/holoviz/panel/releases/tag/v1.1.0
https://github.com/holoviz/panel/releases/tag/v1.1.1

mweinelt commented:
Panel 1.0.0 requries bokeh 3.1.1. Rebased and debugg bokeh build and working on panel build.

mweinelt committed:
fixup! python310Packages.bokeh: 2.4.3 -> 3.2.0

mweinelt committed:
amend! python310Packages.bokeh: 2.4.3 -> 3.2.0

python310Packages.bokeh: 2.4.3 -> 3.2.1

https://docs.bokeh.org/en/latest/docs/releases.html#release-3-2-1

Changes the build to a source build and enables tests.

To enable the tests we have to fetch the sources from Git, which in turn
requires us to build the static files from source.

Bokeh uses setuptools-git-versioning, which has no version override, so
we also fix the broken fallback version in the package metadata.

We skip sampledata tests, because that would require an additional
download, and we skip selenium tests because they are crashing with
chromium/chromedriver and getting out of sync with geckodriver/firefox.

NickCao commented:
Tests are enabled in https://github.com/NixOS/nixpkgs/pull/252993
